rust/noosphere-gateway/src/context_manager.rs

```diff
@@ -0,0 +1,74 @@
+use anyhow::Result;
+use async_stream::try_stream;
+use async_trait::async_trait;
+use noosphere_core::context::HasMutableSphereContext;
+use noosphere_core::data::Did;
+use noosphere_storage::Storage;
+use std::pin::Pin;
+use tokio_stream::Stream;
+
+type ContextManagerSphereStream<'a, C> = dyn Stream<Item = Result<C>> + Send + 'a;
+
+/// Trait to abstract away accessing sphere contexts in the gateway.
+#[async_trait]
+pub trait ContextManager<C, S>: Clone + Send + Sync
+where
+    C: HasMutableSphereContext<S>,
+    S: Storage + 'static,
+{
+    /// Retrieve a sphere context whose counterpart matches `counterpart`.
+    async fn get_sphere_context(&self, counterpart: &Did) -> Result<C>;
+
+    /// Iterate over all managed spheres. Currently used by task workers.
+    /// Will need to rethink this.
+    fn iter(&self) -> Pin<Box<ContextManagerSphereStream<'_, C>>>;
+}
+
+/// Implements [ContextManager] for a single sphere context, used in the single-sphere
+/// gateway workflow.
+#[derive(Clone)]
+pub struct SingleSphereContextManager<C, S>
+where
+    C: HasMutableSphereContext<S>,
+    S: Storage + 'static,
+{
+    context: C,
+    marker: std::marker::PhantomData<S>,
+}
+
+impl<C, S> SingleSphereContextManager<C, S>
+where
+    C: HasMutableSphereContext<S>,
+    S: Storage + 'static,
+{
+    /// Create a new [SingleSphereContextManager], implementing [ContextManager] for a single sphere `context`.
+    pub async fn new(context: C) -> Result<Self> {
+        Ok(SingleSphereContextManager {
+            context,
+            marker: std::marker::PhantomData,
+        })
+    }
+}
+
+#[async_trait]
+impl<C, S> ContextManager<C, S> for SingleSphereContextManager<C, S>
+where
+    C: HasMutableSphereContext<S>,
+    S: Storage + 'static,
+{
+    async fn get_sphere_context(&self, _counterpart: &Did) -> Result<C> {
+        // Our original single sphere gateways in the cluster use a hashed identity
+        // in its hostname, so we won't get the correct counterpart here. Similarly,
+        // a self-hosted orb server may not have a proper subdomain for the sphere,
+        // with a gateway already bound to a single sphere (via [SingleSphereContextManager]).
+        // Always return the only context we have. [GatewayAuthority] ensure requests
+        // are properly authorized.
```

Comment:
Some of this rationale belongs in the main doc for `SingleSphereContextManager`. Also, please dance around the implementation details of our cloud offering; it is preferable for such details to be omitted from this open source project.